### PR TITLE
Pods 8741

### PR DESCRIPTION
**Pull request by jamdade-sharad**
Code changes to create Toggle for LTA April 2024 Changes and Hide Events 2, 6, 7, 8 & 8A

### CODE REVIEW

File: app/models/EventSelection.scala
Comment (PavelVjalicin) [inline app/models/EventSelection.scala:1]: I don't understand why this file exists. We already have EventType enumeration, not sure which one should be kept.
Instead I would either construct RadioItems in the controller or add a EventSelectionService.
optionsFilteredByHideEvents - Should be removed.

options refactored to:

`def options(values:Seq[EventSelection])`

Controller should pass down the events that need to be shown. 
Events should be filtered like this. 
`allEvents.toSet.diff(eventsToRemove)`

**Review by PavelVjalicin — CHANGES_REQUESTED** [review]
This is my review for now. I haven't done full review, don't quite understand what's happening with Formatters at brief glance seems that it might be overly complicated, not sure if all of those things are needed.

File: app/models/EventSelection.scala
Comment (jamdade-sharad) [reply app/models/EventSelection.scala:1]: Incorporated code review comment

File: app/controllers/EventSelectionController.scala
[the remark this answers is not in the document]
Comment (PavelVjalicin) [reply app/controllers/EventSelectionController.scala:1]: Introduce the following into the controller:
`val ltaShowHideToggle = "lta-events-show-hide"`
and replace all use cases as such:
`eventReportingConnector.getFeatureToggle(ltaShowHideToggle)`

[the remark this answers is not in the document]
Comment (jamdade-sharad) [reply app/controllers/EventSelectionController.scala:1]: done

File: app/controllers/EventSelectionController.scala [hunk 1]
```diff
@@ -71,9 +84,25 @@ class EventSelectionController @Inject()(val controllerComponents: MessagesContr
...
+              }
           }
         }
       )
   }
+
+  private def getFilteredOptions(isEnabled: Boolean, ua: UserAnswers)(implicit messages: Messages): Seq[RadioItem] = {
+    val eventsToRemove: Seq[EventSelection] = TaxYear.getSelectedTaxYear(ua).startYear >= "2024" match {
+      case true => Seq(Event2, Event6, Event7, Event8, Event8A)
+      case false => Seq() // TODO: Include new events once developed Seq(Event25, Event26)
```
Comment (ngmjohnstone) [inline app/controllers/EventSelectionController.scala:100]: Event26 is now not going ahead so can amend comment

File: app/controllers/TaxYearController.scala [hunk 2]
```diff
@@ -75,49 +75,62 @@ class TaxYearController @Inject()(val controllerComponents: MessagesControllerCo
...
   private def filteredRadioItems(isEnabled: Boolean)(implicit request: DataRequest[AnyContent]): Seq[RadioItem] = {
-    if(isEnabled) TaxYear.optionsFiltered(taxYear =>  taxYear.startYear >= "2023") else TaxYear.options
+    if (isEnabled) TaxYear.optionsFiltered(taxYear => taxYear.startYear >= "2023") else TaxYear.options
   }
 
   def onPageLoad(waypoints: Waypoints): Action[AnyContent] = (identify andThen getData() andThen requireData).async { implicit request =>
-    val preparedForm = request.userAnswers.get(TaxYearPage).fold(form)(form.fill)
-    eventReportingConnector.getFeatureToggle("event-reporting-tax-year").flatMap { toggleData =>
-      Future.successful(renderPage(preparedForm, waypoints, Ok, toggleData.isEnabled))
+    eventReportingConnector.getFeatureToggle("lta-events-show-hide").flatMap { ltaToggle =>
```
Comment (ngmjohnstone) [inline app/controllers/TaxYearController.scala:82]: Should use the lta toggle val here

File: app/controllers/EventSelectionController.scala [hunk 3]
```diff
@@ -71,9 +84,25 @@ class EventSelectionController @Inject()(val controllerComponents: MessagesContr
...
+              eventReportingConnector.getFeatureToggle(ltaAbolitionShowHideToggle).flatMap { toggleData =>
+                val displayEventList: Seq[RadioItem] = getFilteredOptions(toggleData.isEnabled, request.userAnswers)
+                Future.successful(Ok(view(form, displayEventList, waypoints)))
+              }
           }
         }
       )
   }
+
+  private def getFilteredOptions(isEnabled: Boolean, ua: UserAnswers)(implicit messages: Messages): Seq[RadioItem] = {
```
Comment (ngmjohnstone) [inline app/controllers/EventSelectionController.scala:97]: Does this def cover all the scenarios?

- Toggle on, >=2024
- Toggle on, < 2024
- Toggle off, >= 2024
- Toggle off, < 2024

File: app/controllers/TaxYearController.scala [hunk 4]
```diff
@@ -75,49 +75,62 @@ class TaxYearController @Inject()(val controllerComponents: MessagesControllerCo
...
+              }
             }
           }
-        }
-      )
+        )
+      }
   }
 
+  private def getFormProviderByLtaToggle(isEnabled: Boolean): Form[TaxYear] = {
```
Comment (ngmjohnstone) [inline app/controllers/TaxYearController.scala:129]: Maybe add `TODO` to track removal of this form provider in future

File: app/forms/mappings/Mappings.scala [hunk 5]
```diff
@@ -60,6 +60,11 @@ trait Mappings extends Formatters with Constraints {
                               args: Seq[String] = Seq.empty)(implicit ev: Enumerable[A]): FieldMapping[A] =
     of(enumerableFormatter[A](requiredKey, invalidKey, args))
 
+  protected def enumerable2024[A](requiredKey: String = "error.required",
```
Comment (ngmjohnstone) [inline app/forms/mappings/Mappings.scala:63]: Same as above

**Review by ngmjohnstone — CHANGES_REQUESTED** [review]
Some comments

File: app/controllers/EventSelectionController.scala (hunk 1 above)
Comment (jamdade-sharad) [reply app/controllers/EventSelectionController.scala]: done

File: app/controllers/EventSelectionController.scala (hunk 3 above)
Comment (jamdade-sharad) [reply app/controllers/EventSelectionController.scala:97]: yes, working as expected

File: app/controllers/TaxYearController.scala [hunk 6]
```diff
@@ -75,49 +75,62 @@ class TaxYearController @Inject()(val controllerComponents: MessagesControllerCo
...
-          val originalUserAnswers = request.userAnswers
-          val vd = originalUserAnswers
-            .get(EventReportingOverviewPage).toSeq.flatten.find(_.taxYear == value).flatMap(_.versionDetails)
-          val versionInfo =
-            (vd.map(_.compiledVersionAvailable), vd.map(_.submittedVersionAvailable), vd.map(_.numberOfVersions)) match {
-              case (Some(true), _, Some(versions)) => VersionInfo(versions, Compiled)
-              case (_, Some(true), Some(versions)) => VersionInfo(versions, Submitted)
-              case _ => VersionInfo(1, NotStarted)
-            }
+      eventReportingConnector.getFeatureToggle("lta-events-show-hide").flatMap { ltaToggle =>
```
[the remark this answers is not in the document]
Comment (jamdade-sharad) [reply app/controllers/TaxYearController.scala]: fixed

File: app/controllers/TaxYearController.scala (hunk 4 above)
Comment (jamdade-sharad) [reply app/controllers/TaxYearController.scala:132]: added TODO comment to removal in future

File: app/forms/mappings/Formatters.scala [hunk 7]
```diff
@@ -185,6 +185,28 @@ trait Formatters {
...
+                .map(Right.apply)
+                .getOrElse(Left(Seq(FormError(key, invalidKey, args))))
+            }
+        }
+
+      override def unbind(key: String, value: A): Map[String, String] =
+        baseFormatter.unbind(key, value.toString)
+    }
+
+  private[mappings] def enumerableFormatter2024[A](requiredKey: String, invalidKey: String,
```
[the remark this answers is not in the document]
Comment (jamdade-sharad) [reply app/forms/mappings/Formatters.scala:206]: added TODO comment to track future removal

File: app/forms/mappings/Mappings.scala (hunk 5 above)
Comment (jamdade-sharad) [reply app/forms/mappings/Mappings.scala:65]: added TODO comment to track future removal

**Review by ngmjohnstone — APPROVED** [review]

**Pull request merged**